File: src/main/scala/nl/tudelft/fragments/spoofax/Printer.scala
package nl.tudelft.fragments.spoofax

import com.google.common.collect.Iterables
import nl.tudelft.fragments.Strategy2
import org.metaborg.core.project.SimpleProjectService
import org.spoofax.interpreter.terms.{IStrategoString, IStrategoTerm}

object Printer {
  val s = Strategy2.spoofax

  def printer(languagePath: String) = {
    val languageLocation = s.resourceService.resolve(languagePath)

    val languageComponents = s.discoverLanguages(languageLocation)
    val component = Iterables.get(languageComponents, 0)
    val languageImpl = Iterables.get(component.contributesTo(), 0)

    val projectService = s.injector.getInstance(classOf[SimpleProjectService])
    projectService.create(languageLocation)

    val project = s.projectService.get(languageLocation)
    val context = s.contextService.get(languageLocation, project, languageImpl)
    val runtime = s.strategoRuntimeService.runtime(component, context, false)

    (term: IStrategoTerm) => s.strategoCommon.invoke(runtime, term, "pp-debug").asInstanceOf[IStrategoString]
  }
}

File: org.metaborg.spg.sentence/src/main/java/org/metaborg/spg/sentence/printer/Printer.java
package org.metaborg.spg.sentence.printer;

import org.metaborg.core.MetaborgException;
import org.metaborg.spoofax.core.stratego.IStrategoCommon;
import org.spoofax.interpreter.terms.IStrategoString;
import org.spoofax.interpreter.terms.IStrategoTerm;
import org.strategoxt.HybridInterpreter;

public class Printer {
    private static final String PRETTY_PRINT_STRATEGY = "pp-debug";
    private final IStrategoCommon stratego;
    private final HybridInterpreter interpreter;

    public Printer(IStrategoCommon stratego, HybridInterpreter interpreter) {
        this.stratego = stratego;
        this.interpreter = interpreter;
    }

    public String print(IStrategoTerm term) {
        try {
            IStrategoTerm program = stratego.invoke(interpreter, term, PRETTY_PRINT_STRATEGY);

            if (!(program instanceof IStrategoString)) {
                throw new PrinterRuntimeException("The pretty-printer returned a non-string.");
            }

            return ((IStrategoString) program).stringValue();
        } catch (MetaborgException e) {
            throw new PrinterRuntimeException("Failed to pretty-print term: " + term, e);
        }
    }
}
